src/peripheral/scb.rs

```diff
@@ -400,11 +406,17 @@ impl SCB {
         unsafe { self.invalidate_dcache(cpuid) };
 
         // Now turn on the D-cache
-        // NOTE(unsafe): We have synchronised access by &mut self
-        unsafe { self.ccr.modify(|r| r | SCB_CCR_DC_MASK) };
+        extern "C" {
+            // see asm-v7m.s
+            fn __enable_dcache();
+        }
 
-        crate::asm::dsb();
-        crate::asm::isb();
+        // NOTE(unsafe): Having &mut self means no ISR, for example, is going to
+        // try to alter this register at the same time.
```

Comment:
The safety is provided by the assembly function entering a critical section, so it doesn't matter that we had a unique reference to `self` anyway. Perhaps this comment and the one below could be combined into a single comment that the asm routine contains a critical section so its access to SCB cannot be raced and it contains the required barriers for cache maintenance operation (and same for icache).

In fact, both methods could be made static now (albeit a breaking change) as neither require access to SCB any more.

Reply:
I can merge the comments.

Making the methods static is a design decision not related to the implementation -- it could have been done before, since the asm code is doing the equivalent of `unsafe { &*SCB::ptr() }`. So it comes down to your preference: should any code anywhere be able to turn on caches as "ambient authority," without having been handed an enabling token? It feels inconsistent with the rest of the API, which tends to _really want_ you to have been passed a phantom peripheral representation to do things.

In either case, I'd prefer to get this fix available without bumping major versions, so making the operations static could wait?

Reply:
(For example, the data cache invalidation/cleaning operations are necessary in cases where you have a CPU with caches, and another device, sharing control of a block of memory, e.g. for DMA, if you haven't simply disabled caching for that section of memory. Right now, the only code with access to these operations must have exclusive control over the entire `SCB`. Assuming this was deliberate, the cache enable operations are strictly less common than other maintenance operations, and should probably not be the first or only ones to become `static` if we want to `static` some things.)

Reply:
I would say our current requirement to have the peripheral token exists in order to use Rust's lifetime system to ensure exclusive access to the peripheral in aid of preventing data races in read-modify-write operations, rather than as a means of safety-enforced access control to choose what code can make system modifications, if that makes sense. We have a number of static methods already and in general they're more ergonomic.

I completely take your point that clean and invalidate ops are the ones that really want to be static though (currently I have to unsafely create an SCB to use them since it's not practical to pass the only SCB instance into e.g. an Ethernet driver). Perhaps let's leave these operations as they are for now and consider making the majority of cache control static later.

Currently our master branch already contains breaking changes to our latest release, but I'm happy to issue a patch release on the current 0.6 series containing this and other non-breaking bug fixes.